File: src/third-party/wtf8/wtf8.ml
(**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

(*
 * WTF-8 is a superset of UTF-8 that allows unpaired surrogates.
 *
 * From ES6 6.1.4, "The String Type":
 *
 *   Where ECMAScript operations interpret String values, each element is
 *   interpreted as a single UTF-16 code unit. However, ECMAScript does not
 *   place any restrictions or requirements on the sequence of code units in
 *   a String value, so they may be ill-formed when interpreted as UTF-16 code
 *   unit sequences. Operations that do not interpret String contents treat
 *   them as sequences of undifferentiated 16-bit unsigned integers.
 *
 * If we try to encode these ill-formed code units into UTF-8, we similarly
 * get ill-formed UTF-8. WTF-8 is a fun name for that encoding.
 *
 * https://simonsapin.github.io/wtf-8/
 *)

type codepoint =
  | Point of int
  | Malformed

type 'a folder = 'a -> int -> codepoint -> 'a

(* WTF-8 is a variable length encoding. The first byte in each codepoint
   determines how many other bytes follow. *)
let needed_bytes c =
  if 0x00 <= c && c <= 0x7F then 1 else
  if 0xC2 <= c && c <= 0xDF then 2 else
  if 0xE0 <= c && c <= 0xEF then 3 else
  if 0xF0 <= c && c <= 0xF4 then 4 else
  0

let codepoint s i = function
  | 1 -> Char.code s.[i]
  | 2 ->
    let b0 = Char.code s.[i] in
    let b1 = Char.code s.[i + 1] in
    ((b0 land 0x1F) lsl 6) lor (b1 land 0x3F)
  | 3 ->
    let b0 = Char.code s.[i] in
    let b1 = Char.code s.[i + 1] in
    let b2 = Char.code s.[i + 2] in
    ((b0 land 0x0F) lsl 12) lor
    ((b1 land 0x3F) lsl 6) lor
    (b2 land 0x3F)
  | 4 ->
    let b0 = Char.code s.[i] in
    let b1 = Char.code s.[i + 1] in
    let b2 = Char.code s.[i + 2] in
    let b3 = Char.code s.[i + 3] in
    ((b0 land 0x07) lsl 18) lor
    ((b1 land 0x3F) lsl 12) lor
    ((b2 land 0x3F) lsl 6) lor
    (b3 land 0x3F)
  | _ -> assert false

(* Fold over the WTF-8 code units in a string *)
let fold_wtf_8 ?(pos = 0) ?len f acc s =
  let rec loop acc f s i l =
    if i = l then acc else
    let need = needed_bytes (Char.code s.[i]) in
    if need = 0 then (loop [@tailcall]) (f acc i Malformed) f s (i + 1) l else
    let rem = l - i in
    if rem < need then f acc i Malformed else
    (loop [@tailcall]) (f acc i (Point (codepoint s i need))) f s (i + need) l
  in
  let len = match len with
  | None -> String.length s - pos
  | Some l -> l
  in
  loop acc f (Bytes.unsafe_of_string s) pos len

(* Add a UTF-16 code unit to a buffer, encoded in WTF-8. *)
let add_wtf_8 buf code =
  let w byte = Buffer.add_char buf (Char.unsafe_chr byte) [@@inline] in
  if code >= 0x10000 then begin
  (* 4 bytes *)
    w (0xf0 lor (code lsr 18));
    w (0x80 lor ((code lsr 12) land 0x3F));
    w (0x80 lor ((code lsr 6) land 0x3F));
    w (0x80 lor (code land 0x3F))
  end else if code >= 0x800 then begin
  (* 3 bytes *)
    w (0xe0 lor (code lsr 12));
    w (0x80 lor ((code lsr 6) land 0x3F));
    w (0x80 lor (code land 0x3F))
  end else if code >= 0x80 then begin
  (* 2 bytes *)
    w (0xc0 lor (code lsr 6));
    w (0x80 lor (code land 0x3F))
  end else
  (* 1 byte *)
    w code

File: rust_port/crates/flow_parser/src/wtf8.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! WTF-8 is a superset of UTF-8 that allows unpaired surrogates.
//!
//! From ES6 6.1.4, "The String Type":
//!
//!   Where ECMAScript operations interpret String values, each element is
//!   interpreted as a single UTF-16 code unit. However, ECMAScript does not
//!   place any restrictions or requirements on the sequence of code units in
//!   a String value, so they may be ill-formed when interpreted as UTF-16 code
//!   unit sequences. Operations that do not interpret String contents treat
//!   them as sequences of undifferentiated 16-bit unsigned integers.
//!
//! If we try to encode these ill-formed code units into UTF-8, we similarly
//! get ill-formed UTF-8. WTF-8 is a fun name for that encoding.
//!
//! https://simonsapin.github.io/wtf-8/

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum Codepoint {
    Point(u32),
    Malformed,
}

/// WTF-8 is a variable length encoding. The first byte in each codepoint
/// determines how many other bytes follow.
#[inline]
fn needed_bytes(c: u8) -> usize {
    match c {
        0x00..=0x7F => 1,
        0xC2..=0xDF => 2,
        0xE0..=0xEF => 3,
        0xF0..=0xF4 => 4,
        _ => 0,
    }
}

fn codepoint(s: &[u8], i: usize, need: usize) -> u32 {
    match need {
        1 => s[i] as u32,
        2 => {
            let b0 = s[i] as u32;
            let b1 = s[i + 1] as u32;
            ((b0 & 0x1F) << 6) | (b1 & 0x3F)
        }
        3 => {
            let b0 = s[i] as u32;
            let b1 = s[i + 1] as u32;
            let b2 = s[i + 2] as u32;
            ((b0 & 0x0F) << 12) | ((b1 & 0x3F) << 6) | (b2 & 0x3F)
        }
        4 => {
            let b0 = s[i] as u32;
            let b1 = s[i + 1] as u32;
            let b2 = s[i + 2] as u32;
            let b3 = s[i + 3] as u32;
            ((b0 & 0x07) << 18) | ((b1 & 0x3F) << 12) | ((b2 & 0x3F) << 6) | (b3 & 0x3F)
        }
        _ => unreachable!(),
    }
}

/// Iterate over the WTF-8 code units in a string
pub(super) fn iter_wtf_8<A>(
    bytes: &[u8],
    pos: Option<usize>,
    len: Option<usize>,
    init: A,
    mut f: impl FnMut(&mut A, usize, Codepoint),
) -> A {
    let pos = pos.unwrap_or(0);
    let len = len.unwrap_or(bytes.len() - pos);
    let l = pos + len;
    let mut acc = init;

    let mut i = pos;
    while i < l {
        let need = needed_bytes(bytes[i]);
        if need == 0 {
            f(&mut acc, i, Codepoint::Malformed);
            i += 1;
        } else {
            let rem = l - i;
            if rem < need {
                f(&mut acc, i, Codepoint::Malformed);
                break;
            } else {
                f(&mut acc, i, Codepoint::Point(codepoint(bytes, i, need)));
                i += need;
            }
        }
    }

    acc
}
